### PR TITLE
Add SpotBugs / Find Security Bugs

### PR DESCRIPTION
**Pull request by bdemers**
Sets the default encoding in `GsonDeserializer` to `UTF_8`
Previously this was unset, and used the system default

### CODE REVIEW

**Comment by bdemers** [comment]
Wow @lhazlewood, I had convinced myself 0.11 was already out. 🤦‍♂ 

I changed the note to say that Gson support was added.

As for the test, I added SpotBugs and Find Security Bugs (which is how I found this initially)
Added a few fixes for those issues as well, a few of them are less critical but were simple (and easier to fix than exclude)

Thoughts? (spotbugs changes are in a08c1fba109aa3335f38fd782d2f9c78ff5d7e75)

**Comment by lhazlewood** [comment]
I meant an actual unit test that inspects the `String` instance and ensures the encoding is `UTF-8`.

File: impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java [hunk 1]
```diff
@@ -99,7 +99,9 @@ public DefaultJwtParser() { }
                      CompressionCodecResolver compressionCodecResolver) {
         this.signingKeyResolver = signingKeyResolver;
         this.key = key;
-        this.keyBytes = keyBytes;
+        if (keyBytes != null) {
+            this.keyBytes = keyBytes.clone();
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java]: Why was this added?  Could be a minor performance hit.

File: impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java [hunk 2]
```diff
@@ -188,7 +190,7 @@ public JwtParser setAllowedClockSkewSeconds(long seconds) {
     @Override
     public JwtParser setSigningKey(byte[] key) {
         Assert.notEmpty(key, "signing key cannot be null or empty.");
-        this.keyBytes = key;
+        this.keyBytes = key.clone();
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java]: Same.

File: api/src/main/java/io/jsonwebtoken/lang/Classes.java [hunk 3]
```diff
@@ -193,10 +196,18 @@ public static boolean isAvailable(String fullyQualifiedClassName) {
...
+                @Override
+                public Object run() {
+                    method.setAccessible(true);
+                    return null;
+                }
+            });
+
             return(T)method.invoke(null, args);
-        } catch (Exception e) {
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
```
Comment (lhazlewood) [inline api/src/main/java/io/jsonwebtoken/lang/Classes.java]: This is not backwards-compatible functionality, we can't really do this until 1.0, right?  If any other RuntimeException is thrown, it won't be caught (like it is today).

File: impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParserBuilder.java [hunk 4]
```diff
@@ -138,7 +138,7 @@ public JwtParserBuilder setAllowedClockSkewSeconds(long seconds) {
     @Override
     public JwtParserBuilder setSigningKey(byte[] key) {
         Assert.notEmpty(key, "signing key cannot be null or empty.");
-        this.keyBytes = key;
+        this.keyBytes = key.clone();
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParserBuilder.java]: Is this necessary?

File: impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveProvider.java [hunk 5]
```diff
@@ -140,7 +143,7 @@ public static KeyPair generateKeyPair(String jcaAlgorithmName, String jcaProvide
             ECGenParameterSpec spec = new ECGenParameterSpec(paramSpecCurveName);
             g.initialize(spec, random);
             return g.generateKeyPair();
-        } catch (Exception e) {
+        } catch (InvalidAlgorithmParameterException | NoSuchAlgorithmException | NoSuchProviderException e) {
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveProvider.java]: Same comment about RuntimeExceptions that would be missed but are caught today.

File: impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveSignatureValidator.java [hunk 6]
```diff
@@ -50,7 +50,7 @@ public boolean isValid(byte[] data, byte[] signature) {
              * **/
             byte[] derSignature = expectedSize != signature.length && signature[0] == 0x30 ? signature : EllipticCurveProvider.transcodeSignatureToDER(signature);
             return doVerify(sig, publicKey, data, derSignature);
-        } catch (Exception e) {
+        } catch (InvalidKeyException | java.security.SignatureException e) {
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveSignatureValidator.java]: idem

File: impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java (hunk 1 above)
Comment (bdemers) [reply impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java]: to prevent `keyBytes` from mutating at some point later on. The parser is meant to be reused, so perf _shouldn't_ be an issue, but... YMMV

File: api/src/main/java/io/jsonwebtoken/lang/Classes.java (hunk 3 above)
Comment (bdemers) [reply api/src/main/java/io/jsonwebtoken/lang/Classes.java]: Technically, but I think this one might be a far edge case, a RuntimeException would have to be thrown, and the caller would need to be catching an `IllegalStateException`.

**Comment by bdemers** [comment]
Probably just change the name, but let's wrap up the other open questions/issues first and then figure out the best option

File: api/src/main/java/io/jsonwebtoken/lang/Classes.java (hunk 3 above)
Comment (lhazlewood) [reply api/src/main/java/io/jsonwebtoken/lang/Classes.java]: True - but it is possible given the number of projects that use jjwt.  I'd feel safer if we don't change this until 1.0.

Comment (lhazlewood) [reply api/src/main/java/io/jsonwebtoken/lang/Classes.java]: Could you add a //TODO on the three places and add a follow-up GitHub issue for us to remove it for 1.0 (and add it to the 1.0 release tag?)

**Comment by bdemers** [comment]
@lhazlewood have a suggestion on a good way to do this?
> I meant an actual unit test that inspects the String instance and ensures the encoding is UTF-8.

The only option I could think of was to create a byte array with characters known to render differently, but IMHO, it's easier to scan the source tree for these types of issues 🤔

**Comment by lhazlewood** [comment]
@bdemers just create a mock gson instance and mock `gson.fromJson`, and assert that the argument is UTF-8 compatible?

I found this: https://www.turro.org/publications/?item=114&page=0

But if you don't want to mess with that, no objection from me.  I'm just paranoid 😅

**Comment by bdemers** [comment]
I was messing with creating keys today with BC (via JCA), and it's really really easy to get a `RuntimeException` to be thrown.  I think if we just tack on a `| RuntimeException` we should be good for backward compatibility.

**Comment by lhazlewood** [comment]
> I think if we just tack on a `| RuntimeException` we should be good for backward compatibility.

If that's the case, why change the code at all?  Just curious.

**Comment by bdemers** [comment]
Valid point, I was just thinking it would be easier to change in the future, but it's probably just noise

**Comment by bdemers** [comment]
@lhazlewood back to the topic of the test, we actually do have tests that fail when the system's default encoding is NOT `UTF-8` we just don't see this because that is usually what Linux and MacOS is set to.

Adding a test for this is challenging because it would require forking a JVM to change the `file.encoding` property.

I tried this from my IDE, but you _should_ be able to reproduce it by running the tests against `master` with following arg`-Dfile.encoding=ISO-8859-1`

**Comment by bdemers** [comment]
I updated the PR name to better reflect the changes here, and I've moved the Gson change log note to: #547

**Comment by lhazlewood** [comment]
@bdemers what do you want to do with this? Thoughts?

**Comment by bdemers** [comment]
I'll take another pass at this in a different PR, based off the latest code
